builder/builder-next/controller.go

```diff
@@ -36,12 +38,93 @@ import (
 	"github.com/moby/buildkit/util/archutil"
 	"github.com/moby/buildkit/util/entitlements"
 	"github.com/moby/buildkit/util/leaseutil"
+	"github.com/moby/buildkit/util/network/netproviders"
 	"github.com/moby/buildkit/worker"
+	"github.com/moby/buildkit/worker/containerd"
+	"github.com/moby/buildkit/worker/label"
 	"github.com/pkg/errors"
 	bolt "go.etcd.io/bbolt"
 )
 
-func newController(rt http.RoundTripper, opt Opt) (*control.Controller, error) {
+func newController(rt http.RoundTripper, opt Opt) (*mobycontrol.Controller, error) {
+	if opt.UseSnapshotter {
+		return newSnapshotterController(rt, opt)
+	}
+	return newGraphDriverController(rt, opt)
+}
+
+func newSnapshotterController(rt http.RoundTripper, opt Opt) (*mobycontrol.Controller, error) {
+	if err := os.MkdirAll(opt.Root, 0711); err != nil {
+		return nil, err
+	}
+
+	dist := opt.Dist
+
+	cacheStorage, err := bboltcachestorage.NewStore(filepath.Join(opt.Root, "cache.db"))
+	if err != nil {
+		return nil, err
+	}
+
+	nc := netproviders.Opt{
+		Mode: "auto",
+	}
+	dns := getDNSConfig(opt.DNSConfig)
+
+	wo, err := containerd.NewWorkerOpt(opt.Root, opt.ContainerdAddress, opt.Snapshotter, opt.ContainerdNamespace,
+		opt.Rootless, map[string]string{
+			label.Snapshotter: opt.Snapshotter,
+		}, dns, nc, opt.ApparmorProfile, false, nil, "", ctd.WithTimeout(60*time.Second))
+	if err != nil {
+		return nil, err
+	}
+
+	policy, err := getGCPolicy(opt.BuilderConfig, opt.Root)
+	if err != nil {
+		return nil, err
+	}
+
+	wo.GCPolicy = policy
+	wo.RegistryHosts = opt.RegistryHosts
+
+	w, err := mobyworker.NewContainerdWorker(context.TODO(), wo)
```

Comment:
Is this a "TODO" context? (and if it's "TODO", do we know where this context should come from?

I went looking at the chain of calls that arrive here, and nowhere in the path is a context (none of these take a context as argument);

- -> `main()`
- -> `newDaemonCommand()`
- -> `runDaemon()`
- -> `daemon.start()`
- -> `newRouterOptions()`
- -> `builder-next.New()`
- -> `newController`
    - -> `newSnapshotterController`
    - -> `newGraphDriverController`

So from the above, there's no sensible context to pass (currently).

Looking at what the context is used for, I (at a first glance) saw two uses;

- To set/get the (default) namespace (`withDefaultNamespace()` is used in some locations)
- To get a logger; `NewWorker` (in BuildKit vendor code) uses `bklog.G(ctx).Warnf("git source cannot be enabled: %v", err)`

So (although I'm wondering if we need the logs during startup), perhaps this is a "NODO", and we should use `context.Background()` (e.g.). If we _do_ think we need to pass the context at some point, perhaps we need a comment outlining "why".

(Sorry for the slightly "rambling" comments, mostly put down my notes when I was trying to figure out how/where it's used).

Reply:
This context is also used for the leases manager in buildkit's code.

Also note that in this file there is some `context.TODO`s in `newGraphDriverController` [see here](https://github.com/thaJeztah/docker/blob/c8d_build/builder/builder-next/controller.go#L266-L272)

So basically yes, this is a `TODO` context and it should come from `main` I guess and it should be a context with a timeout I guess.

Reply:
I added a commit for easier review: [`d440891` (#44079)](https://github.com/moby/moby/pull/44079/commits/d4408910abf6a898510751c242e0df889598b050)